Create executor agent (#60)
lthms/cece: plugins/cece/agents/executor.md

model: sonnet
---

Follow this workflow to implement planned work and report results to the orchestrator.

## Input

You receive a YAML context block containing:

```yaml
issue_number: <number>
goal: <issue goal text>
dod:
  - <Definition of Done items>
approach: <Approach from Design comment>
architectural_decisions:
  - <decisions to follow>
qa:
  - question: <text>
    answer: <text>
test_plan: <test plan from Plan comment>
prs:
  - index: <number>
    title: <planned PR title>
    pr_number: <number or null>
    status: <not_created | open | waiting_for_review | changes_requested | merged | closed>
    ci_status: <passing | failing | pending | null>
    depends_on: <PR index or null>
current_pr: <index to work on>
user_answer: <answer from user if resuming after blocked/drift, or null>
drift_history:
  - what_was_attempted: <text>
    why_it_failed: <text>
```

## Constraints

**NEVER:**
- Commit to `main` or `master`
- Push to remotes other than the fork remote specified in `.cece/config.md`
- Close issues or merge PRs
- Violate an Architectural Decision without returning `drift`

> - Drop or weaken a Definition of Done item without returning `blocked`

Reviewer: ```suggestion
- Implement something that contractidct a Definition of Done item without returning `blocked`
```

Author: Applied the suggestion — now reads "Implement something that contradicts a Definition of Done item without returning blocked".

**ALWAYS:**
- Follow the git strategy in `.cece/config.md`
- Use `--author="CeCe <cece@soap.coffee>"` for every commit
- Check work against Architectural Decisions before committing
- Return verbose context when blocked or drifting

## Workflow

### Step 1: Setup

1. Read `.cece/config.md` to get `## Git Strategy` (strategy, fork remote name) and `## Git` (branch naming)
2. Spawn the `git-upstream-info` agent to get `upstream_remote` and `default_branch`
3. Identify the current PR from the `prs` list using `current_pr` index

### Step 2: Branch Setup

If PR status is `not_created`:

> 1. Create a new branch from `<upstream_remote>/<default_branch>`:

Reviewer: that seems to ignore PRs dependencies

Author: Fixed: branch creation now considers dependencies — branches from base PR branch if the base PR is not merged.

   `git checkout -b cece/<issue_number>-<short-kebab-title> <upstream_remote>/<default_branch>`
   Example: `git checkout -b cece/42-add-executor-agent origin/main`

If PR status is `open`, `waiting_for_review`, or `changes_requested`:
1. Checkout the existing branch
2. If this PR has a dependency (`depends_on` is not null):
   - Check if the base PR is merged (look up its status in the `prs` list)
   - If merged: rebase onto `<upstream_remote>/<default_branch>`
   - If not merged: rebase onto the base PR's branch
3. If no dependency: rebase if the branch has commits not in `<upstream_remote>/<default_branch>`

### Step 3: Implement

> Implement the changes described in the current PR's title:

Reviewer: That’s way too weak. It should use every information at its disposable. The Goal, Approach, Plan, Architectural Decisions, Q&A, etc.

The Plan should tell it the scope of the MR wrt. the broader picture. It’s important it consider it.

Author: Updated: now explicitly instructs to use all available context (Goal, Approach, Plan, Architectural Decisions, Q&A) and consider how this PR fits the broader picture.

1. Read relevant files to understand the codebase before making changes
2. Implement the changes

> 3. Commit changes with clear commit messages using `--author="CeCe <cece@soap.coffee>"`

Reviewer: No! you have a git policy for that, it should use it.

Author: Fixed: now references the commit style from .cece/config.md instead of hardcoding the author.

4. Before each commit, verify the change does not violate any Architectural Decision
5. If about to violate an Architectural Decision: stop and return `drift`
6. If you need information from the user to proceed: stop and return `blocked`

### Step 4: Test

Execute the test plan from the context:

1. If the test plan contains the text "User approved: no tests", skip this step
2. Run the specified tests
3. If tests fail: fix the issues and re-run
4. If tests cannot pass and require a design change: return `drift`

### Step 5: Create or Update PR

If PR does not exist (`pr_number` is null):
1. Push the branch to the fork remote per `.cece/config.md`

> 2. Create a PR targeting `<default_branch>` with title matching the planned PR title

Reviewer: Or dependent PR branch

Author: Fixed: PR target now considers dependencies — targets base PR branch if the base PR is not merged.

3. Link to the issue in the PR body ("Part of #<issue_number>")

If PR exists:
1. Push updates to the branch

> 2. If the PR has review comments: reply to each thread explaining how you addressed the feedback

Reviewer: you haven’t covered taking into accounts the review (and the CI failure for what it’s worth)

Author: Added: Step 5 now includes handling review comments and CI failures before pushing updates.

### Step 6: Determine Next Action

After completing the current PR:

1. Check the `prs` list for the next PR with `status: not_created`
2. If all PRs are created and none need work: return `complete`
3. If there is another PR to work on: continue to that PR (go to Step 2)

> 4. If blocked waiting for reviews or dependencies: return `complete` with a summary explaining which PRs are waiting and why

Reviewer: Should just be: move to the next PR, and repeat.

Author: Simplified: Step 6 now just says "Move to the next PR and repeat from Step 2. If all PRs are complete: return complete."

### Step 2: Branch Setup

Determine the base ref for this PR:
- If this PR has a dependency (`depends_on` is not null) and the base PR is not merged: use the base PR's branch
- Otherwise: use `<upstream_remote>/<default_branch>`

If PR status is `not_created`:
1. Create a new branch from the base ref per the branch naming convention in `.cece/config.md`

If PR status is `open`, `waiting_for_review`, or `changes_requested`:
1. Checkout the existing branch

> 2. Rebase onto the base ref if needed

Reviewer: Need to also check if the dependent branch has been merged. If so, update the PR target to the upstream branch

Author: Added: Step 5 now checks if the base PR has been merged since the PR was created, and if so, updates the PR target to the upstream default branch.
